### PR TITLE
Start porting the CWS templates from Tornado to Jinja2

### PR DESCRIPTION
**Pull request by lerks**
This is the first of a series of pull requests that ports CWS's templates from the Tornado engine to Jinja2. This pull request does many atomic changes that gap the syntax differences between the two formats.

---
This change is [Reviewable](https://reviewable.io/reviews/cms-dev/cms/845)

### CODE REVIEW

File: cms/server/contest/templates/documentation.html [hunk 1]
```diff
@@ -8,13 +8,13 @@ <h1>{{ _("Documentation") }}</h1>
 
 <h2>{{ _("Programming languages and libraries") }}</h2>
 
-{% if any(contest.languages|map("get_language")|map(attribute="source_extensions")|select("contains", ".cpp")) %}
+{% if contest.languages|map("get_language")|map(attribute="source_extensions")|select("contains", ".cpp")|list|length > 0 %}
```
Comment (stefano-maggiolo) [inline cms/server/contest/templates/documentation.html]: Can we define our own? like `|nonempty`?

File: cms/server/jinja2_toolbox.py [hunk 2]
```diff
@@ -32,32 +32,64 @@
...
-        args = args[1:]
+        test = ctx.environment.tests[test]
+    for i in l:
+        if not ctx.call(test, i, *args):
+            return False
+    return True
+
+
+@contextfilter
+def any_(ctx, l, test=None, *args):
```
Comment (stefano-maggiolo) [inline cms/server/jinja2_toolbox.py]: Docstring

File: cms/server/jinja2_toolbox.py [hunk 3]
```diff
@@ -32,32 +32,64 @@
...
 from cms.grading.scoretypes import get_score_type
 from cms.grading.tasktypes import get_task_type
 from cms.server.contest.formatting import get_score_class
 
 
-@environmentfilter
-def dictselect(env, d, *args, **kwargs):
-    if len(args) == 0:
+@contextfilter
+def all_(ctx, l, test=None, *args):
```
Comment (stefano-maggiolo) [inline cms/server/jinja2_toolbox.py]: Docstring

File: cms/server/contest/templates/documentation.html (hunk 1 above)
Comment (stefano-maggiolo) [reply cms/server/contest/templates/documentation.html]: I see you defined any later, still nonempty wouldn't be bad for all other occasions you do len() > 0

File: cms/server/contest/jinja2_toolbox.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+from jinja2 import PackageLoader
+
+from cms.server.jinja2_toolbox import GLOBAL_ENVIRONMENT
+
+
+CWS_ENVIRONMENT = GLOBAL_ENVIRONMENT.overlay(
+    extensions=['jinja2.ext.i18n'],
+    loader=PackageLoader('cms.server.contest', 'templates'))
+CWS_ENVIRONMENT.policies['ext.i18n.trimmed'] = True
```
[the remark this answers is not in the document]
Comment (lerks) [reply cms/server/contest/jinja2_toolbox.py]: Done, added comments to every option.

File: cms/server/contest/templates/documentation.html [hunk 5]
```diff
@@ -8,13 +8,13 @@ <h1>{{ _("Documentation") }}</h1>
 
 <h2>{{ _("Programming languages and libraries") }}</h2>
 
-{% if any(".cpp" in get_language(lang).source_extensions for lang in contest.languages) %}
+{% if any(contest.languages|map("get_language")|map(attribute="source_extensions")|select("contains", ".cpp")) %}
```
[the remark this answers is not in the document]
Comment (lerks) [reply cms/server/contest/templates/documentation.html]: Agreed that with get_laguage->to_language and the new any filter this becomes clearer.

File: cms/server/jinja2_toolbox.py [hunk 6]
```diff
@@ -32,7 +32,25 @@
 from future.builtins import *
 from six import iterkeys, itervalues, iteritems
 
-from jinja2 import Environment, StrictUndefined
+from jinja2 import environmentfilter, Environment, StrictUndefined
+
+
+@environmentfilter
+def dictselect(env, d, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (lerks) [reply cms/server/jinja2_toolbox.py]: Done.

File: cms/server/contest/templates/printing.html [hunk 7]
```diff
@@ -44,7 +44,7 @@ <h2 style="margin-bottom: 10px">{{ _("Print") }}</h2>
 
 <h2 style="margin: 40px 0 10px">{{ _("Previous print jobs") }}</h2>
 
-{% set show_date = any(p.timestamp.replace(tzinfo=utc).astimezone(timezone).date() != now.replace(tzinfo=utc).astimezone(timezone).date() for p in printjobs) %}
+{% set show_date = any(printjobs|map(attribute="timestamp")|map("astimezone", timezone)|map(attribute="date")|map("call")|select("ne", (now|astimezone(timezone)).date())) %}
```
[the remark this answers is not in the document]
Comment (lerks) [reply cms/server/contest/templates/printing.html]: Done.

File: cms/server/contest/templates/task_description.html [hunk 8]
```diff
@@ -37,7 +37,7 @@ <h2>{{ _("Statement") }}</h2>
             {{ _("Some suggested translations follow.") }}
         </p>
     {% set task_primary = primary_statements %}
-    {% for statement in sorted(itervalues(task.statements), key=lambda x: x.language_name) %}
+    {% for statement in itervalues(task.statements)|sort(attribute="language_name") %}
```
[the remark this answers is not in the document]
Comment (lerks) [reply cms/server/contest/templates/task_description.html]: Not sure what you mean: `task.statement` is a dict where value `s` has key `s.language`, while dictsort can only sort by plain key or value.

File: cms/server/contest/templates/contest.html [hunk 9]
```diff
@@ -11,7 +11,7 @@
 var LANGUAGES = {
 {% for lang in contest.languages %}
     '{{ lang }}': {
-{% for extension in get_language(lang).source_extensions %}
+{% for extension in (lang|get_language).source_extensions %}
```
[the remark this answers is not in the document]
Comment (lerks) [reply cms/server/contest/templates/contest.html]: Changed to to_language. (PS: it's defined in `cms/server/jinja2_toolbox.py`)

File: cms/server/contest/templates/documentation.html (hunk 1 above)
Comment (lerks) [reply cms/server/contest/templates/documentation.html]: As you mentioned, I added `all` and `any` that make some of these clearer. I think the benefit of `|empty` vs `|length == 0` and `|nonempty` vs `|length > 0` (which, I guess, are brevity and supposed better readability) are not worth the introduction of new filters that may confuse anyone not familiar with our custom idiom.

File: cms/server/jinja2_toolbox.py (hunk 2 above)
Comment (lerks) [reply cms/server/jinja2_toolbox.py]: Done.

File: cms/server/jinja2_toolbox.py (hunk 3 above)
Comment (lerks) [reply cms/server/jinja2_toolbox.py]: Done.

**Comment by stefano-maggiolo** [comment]
LGTM, pending moving those few iteritems(x)|sort to dictsort

**Comment by lerks** [comment]
Fixed up all commits (with rebasing), tested all again, fixed some bugs, and pushed to the `jinja` branch, which will be merged into master as soon as all commits are in.

**Pull request closed without merging**